marlowe-runtime/web-server/Language/Marlowe/Runtime/Web/Server/HistoryClient.hs

```diff
@@ -208,3 +239,106 @@ loadTransactionsClient eb contractId = MarloweSyncClient $ pure clientInit
+          withEvent eb LoadTxContractFound \ev -> do
+            addField ev $ show block
+            pure $ clientIdle version $ SeekTx $ pure (block, createOutput)
+      }
+
+    clientIdle :: MarloweVersion v -> SeekTx v -> ClientStIdle v IO (Maybe TxRecord.TxRecord)
+    clientIdle version seekTx = SendMsgRequestNext ClientStNext
+      { recvMsgRollForward = \_ steps ->
+          withEvent eb LoadTxRollForward \ev -> do
+            seekTx' <- foldlM (stepSeekTx ev) seekTx steps
+            pure case seekTx' of
+              Done prevScriptOutputs' tx _ consumingTx -> SendMsgDone
+                $ Just
+                $ mkTxRecord version (snd $ NE.head prevScriptOutputs') tx
+                $ Just consumingTx
+              _ -> clientIdle version seekTx'
+      , recvMsgRollBackward = \block ->
+          withEvent eb LoadTxRollBackward \ev -> do
+            clientIdle version <$> rollbackStepTx ev block seekTx
+      , recvMsgRollBackCreation =
+          withEvent eb LoadTxRollBackCreation $ const $ pure Nothing
+      , recvMsgWait = do
+          withEvent eb LoadTxWait
+            $ const
+            $ pure
+            $ SendMsgCancel
+            $ SendMsgDone
+            $ case seekTx of
+              Done prevScriptOutputs' tx _ consumingTx ->
+                Just $ mkTxRecord version (snd $ NE.head prevScriptOutputs') tx $ Just consumingTx
+              SeekConsuming prevScriptOutputs' tx ->
+                Just $ mkTxRecord version (snd $ NE.head prevScriptOutputs') tx Nothing
+              _ -> Nothing
+      }
+
+    stepSeekTx
+      :: Event IO r HistoryClientSelector LoadTxRollForwardField
+      -> SeekTx v
+      -> ContractStep v
+      -> IO (SeekTx v)
+    stepSeekTx ev seekTx = \case
+      ApplyTransaction newTx -> case seekTx of
+        SeekTx prevOutputs@(prevOutput :| prevOutputs')
+          | txId == transactionId newTx -> do
+              addField ev $ FoundTx $ show txId
+              pure $ SeekConsuming prevOutputs newTx
+          | otherwise -> pure case scriptOutput $ output newTx of
+              Nothing -> seekTx
+              Just scriptOutput -> SeekTx $ (blockHeader newTx, scriptOutput) :| prevOutput : prevOutputs'
+        SeekConsuming prevOutputs tx -> do
+          addField ev $ FoundConsumer $ show $ transactionId newTx
+          pure $ Done prevOutputs tx (blockHeader newTx) $ transactionId newTx
+        _ -> pure seekTx
+      _ -> pure seekTx
+
+    mkTxRecord
+      :: MarloweVersion v
+      -> TransactionScriptOutput v
+      -> Transaction v
+      -> Maybe TxId
+      -> TxRecord.TxRecord
+    mkTxRecord version input tx consumingTx = TxRecord.TxRecord{..}
+
+    rollbackStepTx
+      :: Event IO r HistoryClientSelector LoadTxRollBackwardField
+      -> Chain.BlockHeader
+      -> SeekTx v
+      -> IO (SeekTx v)
+    rollbackStepTx ev block = \case
+      SeekTx prevOutputs ->pure $ SeekTx $ rollbackPrevOutputs block prevOutputs
```

Comment:
This might be the silliest suggestion I've ever proposed in a PR but here it is XD
```suggestion
      SeekTx prevOutputs -> pure $ SeekTx $ rollbackPrevOutputs block prevOutputs
```